src/cargo/sources/overlay.rs

```diff
@@ -0,0 +1,119 @@
+use std::task::ready;
+
+use crate::sources::IndexSummary;
+
+use super::source::{MaybePackage, Source};
+
+/// A `Source` that overlays a local registry over a remote registry, pretending
+/// that the packages in the local registry are actually available in the remote
+/// registry.
+///
+/// This is a massive footgun and a terrible idea, so we do not (and never will)
+/// expose this publicly. However, it is useful for some very specific private
+/// things, like locally verifying a bunch of packages at a time before any of
+/// them have been published.
+pub struct OverlayConfusionAttack<'gctx> {
+    local: Box<dyn Source + 'gctx>,
+    remote: Box<dyn Source + 'gctx>,
+}
+
+impl<'gctx> OverlayConfusionAttack<'gctx> {
+    pub fn new(local: Box<dyn Source + 'gctx>, remote: Box<dyn Source + 'gctx>) -> Self {
+        Self { local, remote }
+    }
+}
+
+impl<'gctx> Source for OverlayConfusionAttack<'gctx> {
+    fn source_id(&self) -> crate::core::SourceId {
+        self.remote.source_id()
+    }
+
+    fn supports_checksums(&self) -> bool {
+        true
```

Comment:
I would've guessed that this depends on whether both of the sources being overlaid support checksum. Did I miss something?

Reply:
Good point, I'd had in my head (and in the documentation) that the two sources were both `RegistrySource`s. I've updated the docs and this method to reflect that they could be arbitrary `Source`s.